File: linkedlist/160.c
/**
 * tricky part 是遍历的时候 两个head分别走完自己的长度之后再指向对方的head继续走 就可以相遇在intersection 如果没有intersection 则相遇在null
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if(headA == NULL || headB == NULL) return NULL;
    struct ListNode *a, *b;
    a = headA;
    b = headB;
    while(a != b)
    {
        if(a) a = a->next; 
        else a = headB;
        if(b) b = b->next; 
        else b = headA;
    }
    return a;
}

File: linkedlist/160.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *ListNode) *ListNode {
    var dummy *ListNode
    var lena, lenb int 
    dummy = headA
    for dummy != nil {
        lena++
        dummy = dummy.Next
    }
    dummy = headB
    for dummy != nil {
        lenb++
        dummy = dummy.Next
    }
    if lena > lenb {
        for (lena - lenb) > 0 {
            headA = headA.Next
            lena--
        }
    }else {
        for (lenb - lena) > 0 {
            headB = headB.Next
            lenb--
        }        
    }
    for lena > 0 {
        if headA != nil && headA == headB {
            return headA
        }else {
            headA = headA.Next
            headB = headB.Next
        }
        lena--
    }
    return nil
}
